### DISCUSSION TITLE
Support for arbitrary TIFF tag copying, reading, and writing

### DISCUSSION THREAD

**Discussion by antonmarsden** [libvips/libvips#3996] [Ideas]
I've been using an experimental sharp build to make changes to simple GeoTIFF images. Specifically, converting uint16/float32 DEM elevations to terrain rgb or terrarium format. The solution is faster (before tuning) than [rgbify](https://github.com/mapbox/rio-rgbify) and uses 1/10th of the memory. So I am quite happy, except that geotags are not recognised by libvips. It's likely I can copy/paste these tags using exiftools or similar, but it seems like there should be a better way.

Doing a bit more digging, I discovered a few things:
- libtiff doesn't seem to have a solution for handling all unknown tags, but does allow custom [application tags](http://www.libtiff.org/addingtags.html#define-application-tags).
- there have been a number of questions and/or feature requests for reading/writing TIFF tags (not just GeoTIFF) in the past

Now to the question: is enhanced tag support desirable for libvips?

If so, a few approaches could be taken to solve it:
1. Explicitly introduce well-known tags, potentially leading to a lot of maintenance overhead - perhaps why it has not been attempted yet.
2. Allow library users to provide metadata that describes the tags they are interested in capturing and the datatypes they are expecting.
3. Ask the libtiff maintainers to improve tag support. Not sure how this might play out.

Interested in other people's thoughts on this, and whether any of the above approaches are viable.

See also: https://github.com/libvips/libvips/issues/3180

**Comment by jcupitt** [comment]
Hello @antonmarsden,

Yes, it's a general and longstanding problem with the design of libtiff, unfortunately. It would need a new and fancier API to support read and write of unknown tags, and I doubt if the libtiff devs I know are interested.

I think geotiff started using application tags long, long ago, before the scalability problems were obvious. I imagine they'd pick a different solution now if they had their time over.

One way forward would be for geotiff to use to a more flexible format for their metadata. For example, OME-TIFF gets around this problem by putting all their metadata into XML in the `IMAGE_DESCRIPTION` tag. Geotiff could perhaps write their metadata into a set of tags as now (for compatibility), but *also* write a set of tags into `IMAGE_DESCRIPTION` as (perhaps?) JSON. This would make it very simple for third party libraries to manipulate geotiff images. Having two copies of the metadata is a pain, but you could add serial numbers so that readers would always know which metadata set was the most recent.

Another possibility is for geotiff to make a tiny library that just did read and write of their metadata from a TIFF into a big C struct (or perhaps turned it into JSON?). Third party libraries could then just link (or even copy-paste) this very small thing into their code and save a lot of maintenance headaches. Perhaps geotiff already provide something like this? I've not checked.

We probably need to ask a geotiff expert for an opinion.

**Comment by jcupitt** [comment]
Re. adding more TIFF tags to libvips -- we try to keep the number of tags down if we can. As you say, it adds to testing and maintenance effort, so I'd be reluctant to add more code unless the benefits are large. The libvips TIFF load/save operations are hairy enough already.
